### PR TITLE
Fix base64Encode adding trailing bytes on small strings

### PR DESCRIPTION
**Pull request by kmichel-aiven**
**Changelog category:**
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

**Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):**
Fix base64Encode adding trailing bytes on small strings

Sometimes, base64Encode returns data that belongs to the next string in the buffer.
This test, when failing, shows something like that:

    -SELECT * FROM tabl_1 SETTINGS log_comment = ?;
    -SELECT * FROM tabl_2 SETTINGS log_comment = ?;
    +SELECT * FROM tabl_1 SETTINGS log_comment = ?;\0S
    +SELECT * FROM tabl_2 SETTINGS log_comment = ?;\0c

Where the S at the end of the first string is the first S of the next string in the buffer.

I don't know exactly why, but reading from `query_log` is how I manage to reliably trigger the bug.
The same code on literals, or even on a different table, did not trigger the bug, something related to alignment maybe.

I know it's encode and not decode because in my initial discovery the decoding part was outside of ClickHouse.

The "fast" mode of turbo base64 seem to go too fast on small strings, disabling the AVX2 optimisation for small string fixes the problem.

Disabling this optimisation might have visible performance impacts, maybe there is a better way to fix that.

Also, there is something about MSAN near this code, but I don't know anything about MSAN, no idea if it's related:
https://github.com/ClickHouse/ClickHouse/blob/b39c19399ff951a598dd35e575c715c5c4dfb5a2/src/Functions/FunctionBase64Conversion.h#L156-L158

### CODE REVIEW

**Comment by den-crane** [comment]
issue https://github.com/ClickHouse/ClickHouse/issues/30854

**Comment by den-crane** [comment]
```
SELECT substring(hex(base64Decode(base64Encode(repeat('a', 46)))), -10) AS r
┌─r──────────┐
│ 6161610000 │
└────────────┘
```

File: tests/queries/0_stateless/02113_base64encode_trailing_bytes.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+set log_queries=1;
+DROP TABLE system.query_log;
```
Comment (Algunenano) [inline tests/queries/0_stateless/02113_base64encode_trailing_bytes.sql:2]: Please don't do this, this affects the whole cluster (and all tests).

You can use things like `$CLICKHOUSE_DATABASE` and so on to create distinct queries.

**Comment by pkit** [comment]
It looks like source buffer must be 32 bytes longer for short strings, see [here](https://github.com/powturbo/Turbo-Base64/blob/15e044269d747bb30c1692cec921ce6ca2128720/turbob64avx2.c#L167)
But in CH it's exaclty the srclen of the source string: [here](https://github.com/ClickHouse/ClickHouse/blob/b39c19399ff951a598dd35e575c715c5c4dfb5a2/src/Functions/FunctionBase64Conversion.h#L122).

**Comment by kmichel-aiven** [comment]
> It looks like source buffer must be 32 bytes longer for short strings, see [here](https://github.com/powturbo/Turbo-Base64/blob/15e044269d747bb30c1692cec921ce6ca2128720/turbob64avx2.c#L167) But in CH it's exaclty the srclen of the source string: [here](https://github.com/ClickHouse/ClickHouse/blob/b39c19399ff951a598dd35e575c715c5c4dfb5a2/src/Functions/FunctionBase64Conversion.h#L122).

That's probably the issue, they don't say if the data in the extra 32 bytes affects the result or just "should be readable and will be correctly ignored".
If it affects the result, I don't think that constraint is compatibles with "columnar" strings packed one after the other in a big buffer, and padding every string with 32 bytes of null every time doesn't sound like a great idea.
(Unless it's more subtle, and the result will be correct as long as the 32 bytes are readable and properly "owned" by the calling code, even if non-null, but that's beyond what I can understand about C++ and compiler optimizations.)

**Comment by pkit** [comment]
@kmichel-aiven I'm pretty sure it's about 32 byte padding between strings, as it used memory aligned ops with 256 byte stride (AVX2).
I.e. you're right: it will probably never work correctly for packed strings.

**Comment by den-crane** [comment]
Probably it has no sense to fix because of https://github.com/ClickHouse/ClickHouse/issues/31643

**Comment by pkit** [comment]
Dunno, from https://github.com/powturbo/Turbo-Base64#benchmark-incl-the-best-simd-base64-libs
It looks like `no-avx2` `tb64` is still faster than `avx2` `fastbase64`...

**Comment by alexey-milovidov** [comment]
Ok, before we have ditched this library, cutting it also LGTM.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by den-crane** [comment]
There is no sense to use query_log in the test.

simple 

```sql
SELECT substring(hex(base64Decode(base64Encode(repeat('a', 46)))), -10) AS r
┌─r──────────┐
│ 6161610000 │ -- expected 616161
└────────────┘
```

**Pull request merged**